### PR TITLE
fix(git): Fixes Git when using auth or fetch. Fixes #2343

### LINKED ISSUE #2343: git artifact fetch of reference fails due to missing credentials
Checklist:

 * [ x] I've included the version.
 * [ x] I've included reproduction steps.
 * [ x] I've included the workflow YAML.
 * [ x] I've included the logs.
 
**What happened**:
I tried to checkout a gerrit change reference, using git artifacts. It failed due to the fetch didn't use authentication.

**What you expected to happen**:
The repo is cloned and the change is fetched and checked out.

**How to reproduce it (as minimally and precisely as possible)**:
1) Set up a git server with authentication enabled
2) Try to fetch and checkout a reference from that server

**Anything else we need to know?**:
To verify the setup I tried a clone and checkout a revision that exists, like a branch, and then it works, so it's the `fetch` that fails.
It can also be seen in artifacts/git.go that `Auth` is missing from `FetchOptions` (https://github.com/argoproj/argo/blob/master/workflow/artifacts/git/git.go#L106).
 
**Environment**:
- Argo version:
```
Argo 2.4.3
```
- Kubernetes version :
```
K8s v1.14.8
```

**Other debugging information (if applicable)**:
- workflow:
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: git-test-
spec:
  entrypoint: gerrit-change
  templates:
  - name: gerrit-change
    inputs:
      artifacts:
      - name: my-project
        path: /src
        git:
          repo: https://my-gerrit-server.com/projects/my-project.git
          fetch: ["refs/changes/12/12345/123:change"]
          revision: "change"
          usernameSecret:
            name: gerrit-credentials
            key: username
          passwordSecret:
            name: gerrit-credendtials
            key: password
          insecureIgnoreHostKey: true
    container:
      image: golang:1.10
      command: [sh, -c]
      args: ["git status && ls -la"]
      workingDir: /src
```
- executor logs:
```
Loading...
```
- workflow-controller logs:
```
time="2020-03-02T22:08:22Z" level=info msg="Processing workflow" namespace=argo workflow=git-test-knzvf
time="2020-03-02T22:08:23Z" level=info msg="Updating node git-test-knzvf (git-test-knzvf) status Pending -> Error"
time="2020-03-02T22:08:23Z" level=info msg="Updating node git-test-knzvf (git-test-knzvf) message: failed to load artifacts: authentication required"
time="2020-03-02T22:08:23Z" level=info msg="Updated phase Running -> Error" namespace=argo workflow=git-test-knzvf
time="2020-03-02T22:08:23Z" level=info msg="Updated message  -> failed to load artifacts: authentication required" namespace=argo workflow=git-test-knzvf
time="2020-03-02T22:08:23Z" level=info msg="Marking workflow completed" namespace=argo workflow=git-test-knzvf
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Fixes #2343 
See #3246

# Changes

* Write SSH private key to a temp file rather than `~/.ssh/id_rsa` so we don't delete my keys while testing 😡
* Add support for authentication when used with `fetch` or `revision`.

### CODE REVIEW

File: workflow/artifacts/git/git.go [hunk 1]
```diff
@@ -28,130 +27,136 @@ type GitArtifactDriver struct {
...
+		filename, err := filepath.Abs("git-ask-pass.sh")
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		_, err = os.Stat(filename)
+		if os.IsNotExist(err) {
+			err := ioutil.WriteFile(filename, []byte(`#!/bin/sh
+case "$1" in
+Username*) echo "${GIT_USERNAME}" ;;
+Password*) echo "${GIT_PASSWORD}" ;;
```
Comment (simster7) [inline workflow/artifacts/git/git.go:74]: Possible smell: writing passwords to the file system?

File: workflow/artifacts/git/git.go [hunk 2]
```diff
@@ -28,130 +27,136 @@ type GitArtifactDriver struct {
...
 		err = fetchOptions.Validate()
 		if err != nil {
-			return errors.InternalWrapError(err)
+			return err
 		}
-
 		err = repo.Fetch(&fetchOptions)
-		if err != nil {
-			return errors.InternalWrapError(err)
+		if err != nil && err.Error() != "already up-to-date" {
```
Comment (simster7) [inline workflow/artifacts/git/git.go:128]: Extract to function or constant? `IsAlreadyUpToDateErr(err)`

File: workflow/artifacts/git/git.go (hunk 1 above)
Comment (alexec) [reply workflow/artifacts/git/git.go:74]: this is copied from Argo CD, the actual password is not written to a file

File: workflow/artifacts/git/git.go (hunk 2 above)
Comment (alexec) [reply workflow/artifacts/git/git.go:128]: good idea - there was no easy way to check this

Comment (alexec) [reply workflow/artifacts/git/git.go]: done

**Pull request merged**